packages/tokamak/contracts-bedrock/src/tokamak-contracts/verification/L1ContractVerification.sol

```diff
@@ -0,0 +1,351 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/access/Ownable.sol';
+import './interface/IL1ContractVerification.sol';
+import './interface/IProxyAdmin.sol';
+
+contract L1ContractVerification is IL1ContractVerification, Ownable {
+  // ProxyAdmin contract address
+  address public immutable PROXY_ADMIN_ADDRESS;
+
+  struct ChainConfig {
+    mapping(bytes32 => ContractConfig) contractConfigs;
+    SafeConfig safeConfig;
+    address expectedNativeToken;
+    bool safeVerificationRequired;
+  }
+
+  // Consolidated mappings
+  mapping(uint256 => ChainConfig) public chainConfigs;
+
+  // Bridge registry address
+  address public L1BridgeRegistryV1_1Address;
+
+  // Contract IDs
+  bytes32 public constant L1_STANDARD_BRIDGE = keccak256('L1_STANDARD_BRIDGE');
+  bytes32 public constant L1_CROSS_DOMAIN_MESSENGER =
+    keccak256('L1_CROSS_DOMAIN_MESSENGER');
+  bytes32 public constant OPTIMISM_PORTAL = keccak256('OPTIMISM_PORTAL');
+  bytes32 public constant SYSTEM_CONFIG = keccak256('SYSTEM_CONFIG');
+
+  // Events
+  event SafeVerificationRequiredSet(uint256 indexed chainId, bool required);
+
+  constructor(address _proxyAdminAddress) Ownable() {
+    require(_proxyAdminAddress != address(0), 'Invalid proxy admin address');
+    PROXY_ADMIN_ADDRESS = _proxyAdminAddress;
+  }
+
+  // External functions
+  function setContractConfig(
+    bytes32 contractId,
+    bytes32 implementationHash,
+    bytes32 proxyHash,
+    address expectedProxyAdmin
+  ) external onlyOwner {
+    _setContractConfig(
+      block.chainid,
+      contractId,
+      implementationHash,
+      proxyHash,
+      expectedProxyAdmin
+    );
+  }
+
+  function setSafeConfig(
+    address tokamakDAO,
+    address foundation,
+    address thirdOwner,
+    uint256 threshold
+  ) external onlyOwner {
+    _setSafeConfig(
+      block.chainid,
+      tokamakDAO,
+      foundation,
+      thirdOwner,
+      threshold
+    );
+  }
+
+  function setExpectedNativeToken(address tokenAddress) external onlyOwner {
+    require(tokenAddress != address(0), 'Invalid token address');
+    chainConfigs[block.chainid].expectedNativeToken = tokenAddress;
+  }
+
+  function verifyL1Contracts(
+    address systemConfigProxy
+  ) external returns (bool) {
+    require(
+      _verifyL1Contracts(block.chainid, systemConfigProxy),
+      'Contracts verification failed'
+    );
+    emit VerificationSuccess(msg.sender);
+    return true;
+  }
+
+  function verifyAndRegisterRollupConfig(
+    address systemConfigProxy,
+    uint8 _type,
+    address _l2TON,
+    string calldata _name
+  ) external returns (bool) {
+    require(_type == 2, 'Registration allowed only for TON tokens');
+    require(
+      _verifyL1Contracts(block.chainid, systemConfigProxy),
+      'Contracts verification failed'
```

Comment:
When we fail to run require statement, how about writing clearly something like 'SystemConfig verification failed'?